File: Problems/5/main.go
package main

import "fmt"

func main() {
	fmt.Println(longestPalindrome("babad") + "|")
	fmt.Println(longestPalindrome("cbbd") + "|")
	fmt.Println(longestPalindrome(" ") + "|")
	fmt.Println(longestPalindrome("") + "|")
}

func longestPalindrome(s string) string {
	// O(n^2) Solution
	// O(1) Space
	if len(s) == 0 {
		return ""
	}

	maxLen := 0
	mstart, mend := 0, 0

	for i := 1; i < len(s); i++ {
		ts, te := 0, 0
		// aa
		if s[i] == s[i-1] {
			ts, te = i-1, i
			for {
				if ts-1 >= 0 && te+1 < len(s) && s[ts-1] == s[te+1] {
					ts--
					te++
				} else {
					break
				}
			}
			if te-ts+1 > maxLen {
				maxLen = te - ts + 1
				mstart = ts
				mend = te
			}
		}

		// aba
		if i+1 < len(s) && s[i-1] == s[i+1] {
			ts, te = i-1, i+1
			for {
				if ts-1 >= 0 && te+1 < len(s) && s[ts-1] == s[te+1] {
					ts--
					te++
				} else {
					break
				}
			}
			if te-ts+1 > maxLen {
				maxLen = te - ts + 1
				mstart = ts
				mend = te
			}
		}
	}

	return s[mstart : mend+1]
}

File: Problems/5/main.cpp
#include <vector>
#include <iostream>
#include <string>
#include <map>
#include <stack>
#include <algorithm>
#include <numeric>
#include <unordered_set>
#include <unordered_map>
#include <queue>
#include <functional>
#include <sstream>
#include <set>
#include <cctype>

using namespace std;

class Solution
{
public:
    string longestPalindrome(string s)
    {
        int n = s.size();

        int len = 0;
        string res;
        for (int i = 0; i < n; i++)
        {
            // aba
            int l = i - 1, r = i + 1;
            while (l >= 0 && r < n && s[l] == s[r])
            {
                l--;
                r++;
            }
            if (r - l - 1 > len)
            {
                len = r - l - 1;
                res = s.substr(l + 1, len);
            }

            // aa
            l = i - 1;
            r = i;
            while (l >= 0 && r < n && s[l] == s[r])
            {
                l--;
                r++;
            }
            if (r - l - 1 > len)
            {
                len = r - l - 1;
                res = s.substr(l + 1, len);
            }
        }
        return res;
    }
};

// O(n^2) easy solution -> check on each viable spot.
